lib/charms/opensearch/v0/opensearch_tls.py

```diff
@@ -382,3 +412,156 @@ def get_unit_certificates(self) -> Dict[CertType, str]:
+            return None
+
+        stored_certs = run_cmd(f"openssl pkcs12 -in {ca_trust_store} -passin pass:{store_pwd}").out
+
+        # parse output to retrieve the current CA (in case there are many)
+        start_cert_marker = "-----BEGIN CERTIFICATE-----"
+        end_cert_marker = "-----END CERTIFICATE-----"
+        certificates = stored_certs.split(end_cert_marker)
+        for cert in certificates:
+            if f"friendlyName: {alias}" in cert:
+                return f"{start_cert_marker}{cert.split(start_cert_marker)[1]}{end_cert_marker}"
+
+        return None
+
+    def store_new_tls_resources(self, cert_type: CertType, secrets: Dict[str, Any]):
+        """Add key and cert to keystore."""
+        cert_name = cert_type.val
+        if cert_type == CertType.APP_ADMIN:
+            store_pwd = self.charm.secrets.get(Scope.APP, f"keystore-password-{cert_name}")
+        else:
+            store_pwd = self.charm.secrets.get(Scope.UNIT, f"keystore-password-{cert_name}")
+        store_path = f"{self.certs_path}/{cert_type}.p12"
+
+        if not secrets.get("key"):
+            logging.error("TLS key not found, quitting.")
+            return
+
+        # we store the pem format to make it easier for the python requests lib
+        if cert_type == CertType.APP_ADMIN:
+            self.charm.opensearch.write_file(
+                f"{self.certs_path}/chain.pem",
+                secrets["chain"],
+            )
+
+        try:
+            os.remove(store_path)
+        except OSError:
+            pass
+
+        tmp_key = tempfile.NamedTemporaryFile(mode="w+t", suffix=".pem")
+        tmp_key.write(secrets.get("key"))
+        tmp_key.flush()
+        tmp_key.seek(0)
+
+        tmp_cert = tempfile.NamedTemporaryFile(mode="w+t", suffix=".cert")
+        tmp_cert.write(secrets.get("cert"))
+        tmp_cert.flush()
+        tmp_cert.seek(0)
+
+        try:
+            cmd = f"""openssl pkcs12 -export \
+                -in {tmp_cert.name} \
+                -inkey {tmp_key.name} \
+                -out {store_path} \
+                -name {cert_name} \
+                -passout pass:{store_pwd}
+            """
+            if secrets.get("key-password"):
+                cmd = f"{cmd} -passin pass:{secrets.get('key-password')}"
+
+            run_cmd(cmd)
+            run_cmd(f"sudo chmod +r {store_path}")
+        finally:
+            tmp_key.close()
+            tmp_cert.close()
+
+    def all_tls_resources_stored(self, only_unit_resources: bool = False) -> bool:
+        """Check if all TLS resources are stored on disk."""
+        cert_types = [CertType.UNIT_TRANSPORT, CertType.UNIT_HTTP]
+        if not only_unit_resources:
+            cert_types.append(CertType.APP_ADMIN)
+
+        for cert_type in cert_types:
+            if not exists(f"{self.certs_path}/{cert_type}.p12"):
+                # ensure backward compatibility to older charm versions still using pem files
+                for extension in ["key", "cert"]:
```

Comment:
Can you explain this?

Reply:
@Mehdi-Bendriss This addresses an issue I found with the upgrade integration tests (see here: https://github.com/canonical/opensearch-operator/actions/runs/9704294103/job/26784432918). When the integration test deploys and older version of the charm, there are key- and cert-files used instead of keystores and truststores. Now when the test uses `assert_continuous_writes_...`, this gets the admin password via the `get-password` action, which executes this check and failes if there are only p12-files. That's why I added this.

Reply:
We can drop compatibility for pem files starting from this PR, we can luckily afford to do that now that we're still on `2/edge`

Reply:
done